common/src/main/java/com/gregtechceu/gtceu/integration/GTOreVeinWidget.java

```diff
@@ -0,0 +1,138 @@
+
+import com.gregtechceu.gtceu.api.data.worldgen.GTOreDefinition;
+import com.gregtechceu.gtceu.api.data.worldgen.bedrockfluid.BedrockFluidDefinition;
+import com.gregtechceu.gtceu.api.registry.GTRegistries;
+import com.gregtechceu.gtceu.utils.FormattingUtil;
+import com.lowdragmc.lowdraglib.gui.widget.LabelWidget;
+import com.lowdragmc.lowdraglib.gui.widget.SlotWidget;
+import com.lowdragmc.lowdraglib.gui.widget.TankWidget;
+import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup;
+import com.lowdragmc.lowdraglib.jei.IngredientIO;
+import com.lowdragmc.lowdraglib.misc.FluidStorage;
+import com.lowdragmc.lowdraglib.misc.ItemStackTransfer;
+import com.lowdragmc.lowdraglib.side.fluid.FluidStack;
+import com.lowdragmc.lowdraglib.utils.LocalizationUtils;
+import lombok.Getter;
+import net.minecraft.core.NonNullList;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.levelgen.heightproviders.HeightProvider;
+import net.minecraft.world.level.levelgen.heightproviders.UniformHeight;
+import net.minecraft.world.level.material.Fluid;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+/**
+ * @author Arbor
+ * @implNote GTOreVeinWidget
+ */
+@Getter
+public class GTOreVeinWidget extends WidgetGroup {
+    private final String name;
+    private final int weight;
+    private final String range;
+    private final String dimensions;
+    private final Set<ResourceKey<Level>> dimensionFilter;
+
+    public GTOreVeinWidget(GTOreDefinition oreDefinition) {
+        super(0, 0, 120, 160);
+        this.name = getOreName(oreDefinition);
+        this.weight = oreDefinition.getWeight();
+        this.dimensionFilter = oreDefinition.getDimensionFilter();
+        this.dimensions = dimensions();
+        this.range = range(oreDefinition);
+        setClientSideWidget();
+        setupBaseGui(oreDefinition);
+        setupText();
+    }
+
+    public GTOreVeinWidget(BedrockFluidDefinition fluid) {
+        super(0, 0, 120, 140);
+        this.name = getFluidName(fluid);
+        this.weight = fluid.getWeight();
+        this.dimensionFilter = fluid.getDimensionFilter();
+        this.dimensions = dimensions();
+        this.range = "NULL";
+        setClientSideWidget();
+        setupBaseGui(fluid);
+        setupText();
+    }
+
+    private String range(GTOreDefinition oreDefinition) {
+        HeightProvider height = oreDefinition.getRange().height;
+        int maxHeight = height instanceof UniformHeight ? ((UniformHeight) height).maxInclusive.resolveY(null) : 0;
+        int minHeight = height instanceof UniformHeight ? ((UniformHeight) height).minInclusive.resolveY(null) : 0;
```

Comment:
these can be like this:
```suggestion
        int minHeight = 0, maxHeight = 0;
        if (height instanceof UniformHeight uniform) {
            minHeight = uniform.minInclusive.resolveY(null);
            maxHeight = uniform.maxInclusive.resolveY(null);
        }
```